OutOfSchool/OutOfSchool.Redis/MultiLayerCache.cs

```diff
@@ -0,0 +1,125 @@
+    private readonly ICacheService _cacheService;
+    private readonly MemoryCacheConfig memoryCacheConfig;
+
+    public MultiLayerCache(
+        ICacheService externalCache,
+        IMemoryCache inMemoryCache,
+        IOptions<MemoryCacheConfig> memoryCacheConfig
+    )
+    {
+        this._memoryCache = inMemoryCache;
+        this._cacheService = externalCache;
+        this.memoryCacheConfig = memoryCacheConfig.Value;
+    }
+
+    public Task ClearCacheAsync(string key)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+            try
+            {
+                _memoryCache.Remove(key);
+                _cacheService.ClearCacheAsync(key);
+            }
+            finally
+            {
+                cacheLock.ExitWriteLock();
+            }
+        });
+    }
+
+    public async Task<T> GetOrAddAsync<T>(string key, Func<Task<T>> newValueFactory, TimeSpan? absoluteExpirationRelativeToNowInterval = null, TimeSpan? slidingExpirationInterval = null)
+    {
+        T returnValue = default;
+
+        await ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterReadLock();
+            
+            try
+            {
+                if (_memoryCache.TryGetValue(key, out T value))
+                {
+                    returnValue = value;
+                }
+            }
+            finally
+            {
+                cacheLock.ExitReadLock();
+            }
+        });
+
+        if (EqualityComparer<T>.Default.Equals(returnValue, default))
+        {
+            returnValue = await _cacheService.GetOrAddAsync(key, newValueFactory, absoluteExpirationRelativeToNowInterval, slidingExpirationInterval);
+            await SetAsync(key, returnValue, absoluteExpirationRelativeToNowInterval, slidingExpirationInterval);
+        }
+
+        return returnValue;
+    }
+
+    public Task RefreshAsync(string key)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+            try
+            {
+                _cacheService.RefreshAsync(key);
+            }
+            finally
+            {
+                cacheLock.ExitWriteLock();
+            }
+        });
+    }
+
+    public Task SetAsync<T>(string key, T value, TimeSpan? absoluteExpirationRelativeToNowInterval = null, TimeSpan? slidingExpirationInterval = null)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+
+            try
+            {
+                var cacheEntryOptions = new MemoryCacheEntryOptions()
+                {
+                    SlidingExpiration = slidingExpirationInterval ?? memoryCacheConfig.SlidingExpirationInterval,
+                    AbsoluteExpirationRelativeToNow = absoluteExpirationRelativeToNowInterval ?? memoryCacheConfig.AbsoluteExpirationRelativeToNowInterval,
+                    Priority = CacheItemPriority.Normal,
+                    Size = 1024,
```

Comment:
(Magic literals) What `1024` means here? 

Also, accordingly to [SO: What do the size settings for MemoryCache mean?](https://stackoverflow.com/questions/59774566/what-do-the-size-settings-for-memorycache-mean) you could use any number but this `Size` parameter is related to `SizeLimit` parameter of the whole cache. If latter is not set, the former does nothing.

Probably it could be removed at all

Reply:
Added Size to cache config.

Reply:
That's cool, but I asked about possible removing it at all, as it does nothing until you set the `SizeLimit` per whole cache.

Reply:
I  meant that I added Size to config and changed it to whole cache SizeLimit. And deleted it from options of "set" operation.

Reply:
> I meant that I added Size to config and changed it to whole cache SizeLimit. And deleted it from options of "set" operation.

I'm sorry, but I don't see it in a cache config. 
Please either name the file with changes or update your comments to be relevant to code changes

Reply:
I deleted it and forgot to change my comment. Sorry.